substrate/frame/collective/src/lib.rs

```diff
@@ -979,69 +1003,35 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 	/// * The prime account must be a member of the collective.
 	#[cfg(any(feature = "try-runtime", test))]
 	fn do_try_state() -> Result<(), TryRuntimeError> {
-		Self::proposals()
-			.into_iter()
-			.try_for_each(|proposal| -> Result<(), TryRuntimeError> {
-				ensure!(
-					Self::proposal_of(proposal).is_some(),
-					"Proposal hash from `Proposals` is not found inside the `ProposalOf` mapping."
-				);
-				Ok(())
-			})?;
-
 		ensure!(
-			Self::proposals().into_iter().count() <= Self::proposal_count() as usize,
+			Voting::<T, I>::count() <= Self::proposal_count(),
 			"The actual number of proposals is greater than `ProposalCount`"
 		);
-		ensure!(
-			Self::proposals().into_iter().count() == <ProposalOf<T, I>>::iter_keys().count(),
-			"Proposal count inside `Proposals` is not equal to the proposal count in `ProposalOf`"
-		);
 
-		Self::proposals()
-			.into_iter()
-			.try_for_each(|proposal| -> Result<(), TryRuntimeError> {
-				if let Some(votes) = Self::voting(proposal) {
-					let ayes = votes.ayes.len();
-					let nays = votes.nays.len();
+		let mut proposal_indices =
+			sp_runtime::BoundedBTreeSet::<ProposalIndex, T::MaxProposals>::new();
 
-					ensure!(
-						ayes.saturating_add(nays) <= T::MaxMembers::get() as usize,
-						"The sum of ayes and nays is greater than `MaxMembers`"
-					);
-				}
-				Ok(())
-			})?;
-
-		let mut proposal_indices = vec![];
-		Self::proposals()
-			.into_iter()
-			.try_for_each(|proposal| -> Result<(), TryRuntimeError> {
-				if let Some(votes) = Self::voting(proposal) {
-					let proposal_index = votes.index;
-					ensure!(
-						!proposal_indices.contains(&proposal_index),
-						"The proposal index is not unique."
-					);
-					proposal_indices.push(proposal_index);
-				}
-				Ok(())
-			})?;
-
-		<Voting<T, I>>::iter_keys().try_for_each(
-			|proposal_hash| -> Result<(), TryRuntimeError> {
-				ensure!(
-					Self::proposals().contains(&proposal_hash),
-					"`Proposals` doesn't contain the proposal hash from the `Voting` storage map."
-				);
-				Ok(())
-			},
-		)?;
+		for (proposal_hash, votes) in Voting::<T, I>::iter() {
+			ensure!(
+				T::Preimages::is_requested(&proposal_hash),
+				"Preimage for hash contained in `Voting` is not stored."
+			);
 
-		ensure!(
-			Self::members().len() <= T::MaxMembers::get() as usize,
-			"The member count is greater than `MaxMembers`."
-		);
+			let ayes = votes.ayes.len();
+			let nays = votes.nays.len();
+
+			ensure!(
+				ayes.saturating_add(nays) <= T::MaxMembers::get() as usize,
+				"The sum of ayes and nays is greater than `MaxMembers`"
+			);
+
+			let proposal_index = votes.index;
+			ensure!(
+				// returns true if the value was not already in the set
+				proposal_indices.try_insert(proposal_index).unwrap(),
```

Comment:
Unwrap?? 🙈

Reply:
I should have put an expect 😓  it's safe though:
`let mut proposal_indices = sp_runtime::BoundedBTreeSet::<ProposalIndex, T::MaxProposals>::new();`
and we know that there's always at most `T::MaxProposals` active proposals.

Should I Err here too instead?

Reply:
I'd forgotten about this method we have `.defensive`. Let's use that instead of `.unwrap`, if we are 100% sure it is safe.

Reply:
`.defensive_proof` is better as we can clearly document the reasons why this should not have any errors.

Reply:
@KiChjang thanks, but I used `defensive_map_err`, as I think it fits this case better. We're in `do_try_state`, so if the error should ever happen (probably not, but it would definitely be due to an inconsistent state), then it's probably better to return the error instead of going forward.